SML/SML/Matrix/Matrix.cs

```diff
@@ -0,0 +1,312 @@
+
+        Matrix result = new(matrix1.Rows, matrix1.Columns);
+
+        for (int i = 0; i < result.Rows; i++)
+        {
+            for (int j = 0; j < result.Columns; j++)
+            {
+                result[i, j] = matrix1[i, j] - matrix2[i, j];
+            }
+        }
+
+        return result;
+    }
+
+    private static void CheckForAddOrSubOperations(Matrix matrix1, Matrix matrix2)
+    {
+        CheckNullMatrix(matrix1);
+        CheckNullMatrix(matrix2);
+        CheckSameDimensionsMatrix(matrix1, matrix2);
+    }
+
+    public static Matrix operator *(Matrix matrix1, Matrix matrix2)
+    {
+        CheckForMultiplyOperation(matrix1, matrix2);
+
+        Matrix result = new(matrix1.Rows, matrix2.Columns);
+
+        for (int i = 0; i < result.Rows; i++)
+        {
+            for (int j = 0; j < result.Columns; j++)
+            {
+                for (int k = 0; k < matrix2.Rows; k++)
+                {
+                    result[i, j] += matrix1[i, k] * matrix2[k, j];
+                }
+            }
+        }
+
+        return result;
+    }
+
+    private static void CheckForMultiplyOperation(Matrix matrix1, Matrix matrix2)
+    {
+        CheckNullMatrix(matrix1);
+        CheckNullMatrix(matrix2);
+        CheckMultiplyAvailability(matrix1, matrix2);
+    }
+
+    #endregion Operators
+
+    #region Asserts
+
+    private static void CheckMatrixDimensions(int rows, int columns)
+    {
+        CheckRowsNum(rows);
+        CheckColumnsNum(columns);
+    }
+
+    private static void CheckRowsNum(int rows)
+    {
+        if (rows < 0)
+        {
+            throw new ArgumentOutOfRangeException
+                (RowsLessThanZeroExceptionMessage);
+        }
+    }
+
+    private static void CheckColumnsNum(int columns)
+    {
+        if (columns < 0)
+        {
+            throw new ArgumentOutOfRangeException
+                (ColumnsLessThanZeroExceptionMessage);
+        }
+    }
+
+    private void CheckSquareMatrix(Matrix matrix)
+    {
+        if (Rows != matrix.Rows ||
+            Columns != matrix.Columns)
+        {
+            throw new MatrixException(MatricesDifferentDimensionsExceptionMessage);
+        }
+    }
+
+    private static void CheckSameDimensionsMatrix(Matrix matrix1, Matrix matrix2)
+    {
+        if (matrix1.Rows != matrix2.Rows ||
+            matrix1.Columns != matrix2.Columns)
+        {
+            throw new MatrixException(MatricesDifferentDimensionsExceptionMessage);
+        }
+    }
+
+    private static void CheckNullMatrix(Matrix matrix)
+    {
+        if (matrix == null)
+        {
+            throw new ArgumentNullException(nameof(matrix),
+                NullMatrixExceptionMessage);
+        }
+    }
+
+    private static void CheckMultiplyAvailability(Matrix matrix1, Matrix matrix2)
+    {
+        if (matrix1.Columns != matrix2.Rows)
+        {
+            throw new MatrixException(MatricesDifferentDimensionsExceptionMessage);
+        }
+    }
```

Comment:
```suggestion
    private static void CheckMultiplyAvailability(Matrix FirstMatrix, Matrix SecondMatrix)
    {
        if (FirstMatrix.Columns != SecondMatrix.Rows)
        {
            throw new MatrixException(MatricesDifferentDimensionsExceptionMessage);
        }
    }
```

Reply:
Well, Right now I'm thinking about correct name style. For example, "matrix1" has more clear and shorter name, so You can read it faster. Yet "firstMatrix" looks more habitual and correct. So I dunno which one to use

Reply:
But probably names should be consistent, so I'll change it